### ISSUE TITLE
`Vec` and `Box` documentation do not state maximum allocation size guarantee

### ISSUE THREAD

**Issue by aticu** [rust-lang/rust#68593]
In the documentation for [`pointer::add`](https://doc.rust-lang.org/nightly/std/primitive.pointer.html#method.add), it is mentioned that

> `Vec` and `Box` ensure they never allocate more than `isize::MAX` bytes

however, this is not stated in [documentation of `Vec`](https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html), the [module documentation of `vec`](https://doc.rust-lang.org/nightly/alloc/vec/index.html), the [documentation of `Box`](https://doc.rust-lang.org/nightly/alloc/boxed/struct.Box.html) or the [module documentation of `boxed`](https://doc.rust-lang.org/nightly/alloc/boxed/index.html).

This can cause doubt, whether this statement actually holds true, since the places where one would expect such guarantees to be made (especially the "Guarantees" section of the `Vec` documentation) does not mention these guarantees.

**Comment by cuviper** [comment]
Along these lines, I want to point out that this a real system limit now, beyond just what Rust wants to do, at least if you're using the system allocator. This is mentioned in the [glibc 2.30 release notes](https://sourceware.org/ml/libc-alpha/2019-08/msg00029.html):

```
* Memory allocation functions malloc, calloc, realloc, reallocarray, valloc,
  pvalloc, memalign, and posix_memalign fail now with total object size
  larger than PTRDIFF_MAX.  This is to avoid potential undefined behavior with
  pointer subtraction within the allocated object, where results might
  overflow the ptrdiff_t type.
```

C and C++ `ptrdiff_t` is essentially the same as Rust's `isize`, so `PTRDIFF_MAX == isize::MAX`.

**Comment by aticu** [comment]
> I want to point out that this a real system limit now, beyond just what Rust wants to do, at least if you're using the system allocator.

If that's the case, it might make sense to document that constraint on every allocation based API in the standard library, since it should apply to all of them.

**Comment by poliorcetics** [comment]
This has been fixed by #68711 and could be closed
